### PR TITLE
Add new fpm subproject `icar-trainer`

### PR DESCRIPTION
**Pull request by ktras**
Add new fpm subproject, icar-trainer, which removes dependencies from inference-engine.

### CODE REVIEW

**Comment by rouson** [comment]
@ktras I'm getting a seg fault with the trailing output below when I run `cloud-microphysics/setup.sh`.   Presumably that's why this is a Draft PR.  Let me know when it's ready for review.

```
icar-trainer-test                      done.
[100%] Project compiled successfully.
 
A NetCDF_file_t object
icar-trainer-test(61588,0x1d814dec0) malloc: *** error for object 0x726f727265206f4e: pointer being freed was not allocated
icar-trainer-test(61588,0x1d814dec0) malloc: *** set a breakpoint in malloc_error_break to debug

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x102c17537
#1  0x102c16513
#2  0x181fcba23
#3  0x181f9ccbf
#4  0x181eaca3f
#5  0x181dc3b07
#6  0x181dc73f3
#7  0x181ddbebb
#8  0x102084c7f
#9  0x102090317
#10  0x102084467
#11  0x10216f55f
<ERROR> Execution for object " icar-trainer-test " returned exit code  6
<ERROR> *cmd_run*:stopping due to failed executions
STOP 6
```

**Comment by ktras** [comment]
@rouson Thanks for letting me know. I do not get that segfault though. Can you tell me what compiler and version you are using?

**Comment by rouson** [comment]
In another branch, I'll edit the README.md to indicate that gfortran 13 is the minimum requirement (because gfortran 12 seg faults).

File: cloud-microphysics/test/main.f90 [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+! Terms of use are as specified in LICENSE.txt
+program main
+  implicit none
+
+  real t_start, t_finish
+
+  integer :: passes=0, tests=0
+
+  call cpu_time(t_start)
+#ifndef __INTEL_FORTRAN
```
Comment (rouson) [inline cloud-microphysics/test/main.f90:11]:
```suggestion
```

File: cloud-microphysics/test/main.f90 [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  block 
+    use netCDF_file_test_m, only : netCDF_file_test_t
+    type(netCDF_file_test_t) netCDF_file_test
+    call netCDF_file_test%report(passes, tests)
+  end block
+#endif // __INTEL_FORTRAN
```
Comment (rouson) [inline cloud-microphysics/test/main.f90:17]:
```suggestion
```

File: cloud-microphysics/scripts/run-fpm.sh-header
Comment (rouson) [inline cloud-microphysics/scripts/run-fpm.sh-header:1]: @ktras  Let's delete this file and its encompassing directory.

**Review by rouson — COMMENTED** [review]
@ktras I made a few more comments

File: cloud-microphysics/scripts/run-fpm.sh-header
Comment (ktras) [reply cloud-microphysics/scripts/run-fpm.sh-header:1]: Done.

File: setup.sh [hunk 3]
```diff
@@ -39,94 +37,40 @@ done
...
+  echo "This script assumes usage of gfortran to compile and build"
+  echo "When using other Fortran compilers, please ensure you have fpm downloaded and in your path"
+  echo "Then use the following command to install inference-engine: fpm build --flag \"-fcoarray=single\"."
+  echo "Please replace the coarray flag with appropriate coarray flag for your compiler"
+  exit 1
+fi
+
+if ! command -v fpm > /dev/null ; then
+  if ! command -v brew > /dev/null ; then
+    if ! command -v curl > /dev/null ; then
```
Comment (rouson) [inline setup.sh:50]: The logic I recommend is
```
if (.not. fpm_present)  then
   if (.not. homebrew_present) then
      if (.not. gfortran_present) then
         permission = ask_to_brew_install_fpm
         if (permission) then
           ! insert homebrew installation command
        else 
           error stop "please install fpm and restart"
        end if
     else ! gfortran_present == .true.
        download fpm_source_file.F90
        gfortran fpm_source_file.F90
     end if
  else ! homebrew_present == .true.
      `brew install fpm`
  end if
end if
fpm test (which automatically uses `-fcoarray=single`)
```

File: cloud-microphysics/test/main.f90 [hunk 4]
```diff
@@ -0,0 +1,25 @@
+! Copyright (c), The Regents of the University of California
+! Terms of use are as specified in LICENSE.txt
+program main
```
Comment (rouson) [inline cloud-microphysics/test/main.f90:3]:
```suggestion
program main
  use netCDF_file_test_m, only : netCDF_file_test_t
```

File: cloud-microphysics/test/main.f90 [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  implicit none
+
+  real t_start, t_finish
+
+  integer :: passes=0, tests=0
+
+  call cpu_time(t_start)
+  block 
+    use netCDF_file_test_m, only : netCDF_file_test_t
```
Comment (rouson) [inline cloud-microphysics/test/main.f90:12]: @ktras to play it safe, let's keep all `use` statements at the top of the outermost program unit.
```suggestion
```

File: setup.sh [hunk 6]
```diff
@@ -39,100 +39,61 @@ done
...
 echo ""
-echo "____________________ Inference-Engine has been set up! _______________________" 
+echo "____________________ Inference-Engine has been set up! _______________________"
 echo ""
 echo "To run one of the programs in the example subdirectory, enter a command of the"
 echo "following form at a shell command prompt after replacing <example-base-name>"
 echo "with the base name of a file in the example/ subdirectory:"
 echo ""
-echo "./build/run-fpm.sh run --example <example-base-name>"
+echo "fpm run --example <example-base-name>"
```
Comment (rouson) [inline setup.sh:99]: I'm thinking we should start adding `--profile release` to get optimized builds.  
```suggestion
echo "fpm run --example <example-base-name> --profile release"
```
Let's add similar text in the README.md.  That can be in a new PR.

File: setup.sh [hunk 7]
```diff
@@ -39,100 +39,61 @@ done
 
 set -u # error on use of undefined variable
 
-if ! command -v brew > /dev/null ; then
+if ! command -v gfortran > /dev/null ; then
+  echo "This script assumes usage of gfortran to compile and build"
+  echo "When using other Fortran compilers, please ensure you have fpm downloaded and in your path"
+  echo "Then use the following command to install inference-engine: fpm build --flag \"-fcoarray=single\"."
```
Comment (rouson) [inline setup.sh:45]:
```suggestion
  echo "Then use the following command to install inference-engine: fpm build --flag \"<compiler-specific-coarray-flag>\"."
```

File: setup.sh [hunk 8]
```diff
@@ -39,100 +39,61 @@ done
...
+  echo "Please replace the coarray flag with appropriate coarray flag for your compiler"
```
Comment (rouson) [inline setup.sh:46]:
```suggestion
  echo "Please replace the text from the opening angular bracket to the ending bracket  with appropriate flag, if any, that your compiler requires to support Fortran's parallel features (e.g., -fcoarray=single for single-image builds with gfortran)."
```

File: setup.sh [hunk 9]
```diff
@@ -39,100 +39,61 @@ done
...
+  exit 1
+fi
+
+install_fpm_from_source()
+{
+  echo "Installing fpm in following location: $PREFIX/bin"
+  echo "Ensure $PREFIX/bin is in your path or rerun script with the --prefix=PREFIX flag"
   if ! command -v curl > /dev/null ; then
+    echo "Need curl to download source file for fpm to install it"
```
Comment (rouson) [inline setup.sh:55]:
```suggestion
    echo "This script sues curl to download source file for fpm to install it"
```

File: setup.sh [hunk 10]
```diff
@@ -39,100 +39,61 @@ done
...
     exit 1
   fi
-  /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
-  if [ $(uname) = "Linux" ]; then
-    if [ -z "$PATH" ]; then
-      PATH=/home/linuxbrew/.linuxbrew/bin/
-    else
-      PATH=/home/linuxbrew/.linuxbrew/bin/:"$PATH"
+  mkdir temp-dir-to-build-fpm-for-inference-engine-installation
+  curl -L -o temp-dir-to-build-fpm-for-inference-engine-installation/fpm.F90 https://github.com/fortran-lang/fpm/releases/download/current/fpm.F90
```
Comment (rouson) [inline setup.sh:60]: Great directory name. 😄

File: cloud-microphysics/setup.sh
Comment (rouson) [inline cloud-microphysics/setup.sh:1]: @ktras is this script tested in CI?  If not, please create an issue for that.  We can talk about how to write a simple test.

**Review by rouson — APPROVED** [review]
LGTM

File: setup.sh (hunk 7 above)
Comment (ktras) [reply setup.sh]: This if clause was unintentionally left in. Based on your reminder yesterday that `fpm` automatically provides the correct coarray flag for each Fortran compiler, this clause is not necessary and I have removed it.

File: setup.sh (hunk 8 above)
Comment (ktras) [reply setup.sh]: This if clause was unintentionally left in. Based on your reminder yesterday that `fpm` automatically provides the correct coarray flag for each Fortran compiler, this clause is not necessary and I have removed it.

File: cloud-microphysics/setup.sh
Comment (ktras) [reply cloud-microphysics/setup.sh:1]: No, it is not. I have created an issue for this.

**Comment by rouson** [comment]
@ktras I think this has been completed and merged into main so I'm closing this issue.  Feel free to reopen it if you think it's still an unresolved issue.

**Comment by rouson** [comment]
@ktras oops... this was already closed. :)

**Pull request merged**